lightning/src/ln/channel.rs

```diff
@@ -1488,32 +1478,67 @@ impl<SP: Deref> Channel<SP> where
 	where
 		L::Target: Logger
 	{
-		let phase = core::mem::replace(&mut self.phase, ChannelPhase::Undefined);
-		let result = if let ChannelPhase::UnfundedV2(chan) = phase {
+		if let ChannelPhase::UnfundedV2(chan) = &mut self.phase {
 			let logger = WithChannelContext::from(logger, &chan.context, None);
-			match chan.funding_tx_constructed(signing_session, &&logger) {
-				Ok((chan, commitment_signed, event)) => {
-					self.phase = ChannelPhase::Funded(chan);
-					Ok((commitment_signed, event))
-				},
-				Err((chan, e)) => {
-					self.phase = ChannelPhase::UnfundedV2(chan);
-					Err(e)
-				},
-			}
+			chan.funding_tx_constructed(signing_session, &&logger)
 		} else {
-			self.phase = phase;
 			Err(ChannelError::Warn("Got a tx_complete message with no interactive transaction construction expected or in-progress".to_owned()))
-		};
-
-		debug_assert!(!matches!(self.phase, ChannelPhase::Undefined));
-		result
+		}
 	}
 
 	pub fn force_shutdown(&mut self, should_broadcast: bool, closure_reason: ClosureReason) -> ShutdownResult {
 		let (funding, context) = self.funding_and_context_mut();
 		context.force_shutdown(funding, should_broadcast, closure_reason)
 	}
+
+	pub fn commitment_signed<L: Deref>(
+		&mut self, msg: &msgs::CommitmentSigned, best_block: BestBlock, signer_provider: &SP, logger: &L
+	) -> Result<(Option<ChannelMonitor<<SP::Target as SignerProvider>::EcdsaSigner>>, Option<ChannelMonitorUpdate>), ChannelError>
+	where
+		L::Target: Logger
+	{
+		let phase = core::mem::replace(&mut self.phase, ChannelPhase::Undefined);
+		match phase {
+			ChannelPhase::UnfundedV2(chan) => {
+				let holder_commitment_point = match chan.unfunded_context.holder_commitment_point {
+					Some(point) => point,
+					None => {
+						let channel_id = chan.context.channel_id();
+						// TODO(dual_funding): Add async signing support.
+						return Err( ChannelError::close(
+							format!("Expected to have holder commitment points available upon finishing interactive tx construction for channel {}",
+								channel_id)));
+					}
+				};
+				let mut funded_channel = FundedChannel {
+					funding: chan.funding,
+					context: chan.context,
+					interactive_tx_signing_session: chan.interactive_tx_signing_session,
+					holder_commitment_point,
+					is_v2_established: true,
+				};
+				let res = funded_channel.commitment_signed_initial_v2(msg, best_block, signer_provider, logger)
```

Comment:
Probably no need to hold up this PR, but is there a reason why `commitment_signed_initial_v2` is a method on `FundedChannel` rather than `PendingV2Channel`? Seems that would be consistent with v1 channels calling `initial_commitment_signed` from an unfunded phase, only transitioning to `FundedChannel` after it returns.